### PR TITLE
logical: use implicit txns for each INSERT and DELETE

### PR DESCRIPTION
**Pull request by yuzefovich**
**logical: use implicit txns for each INSERT and DELETE**

This commit introduces another strategy for processing rows. Previously, we would handle a batch of rows within an explicit txn, and this commit makes it so that each row is handled in a separate implicit txn. This strategy is now used by default and can be disabled via newly-added `logical_replication.consumer.use_implicit_txns.enabled` cluster setting. Usage of the implicit txns allows us to hit 1PC down in the KV layer in some cases.

**logical: remove some overhead with explicit txns**

Previously, when using the explicit txns to perform INSERT and DELETE ingestion queries we would use "vanilla" `Txn` method. This method creates a new internal session data on each invocation of the internal executor methods. This commit avoids doing that by supplying the Txn method with the copy of the session data from the session that started the replication stream - that copy will be adjusted to apply some internal executor overrides but otherwise will be unchanged. In other words, even though this commit is not exactly a no-op, it is effectively one.

Epic: CRDB-39063.

Release note: None

### CODE REVIEW

**Comment by yuzefovich** [comment]
First commit is #126051.

The benchmarks show some slowdown because we now create more state on each `ExecParsed` call.
```
name                                 old time/op    new time/op    delta
LastWriteWinsInsert/noPrevValue-8       802µs ± 1%     830µs ± 1%  +3.55%  (p=0.000 n=9+8)
LastWriteWinsInsert/withPrevValue-8    1.18ms ± 1%    1.21ms ± 1%  +2.64%  (p=0.000 n=10+10)

name                                 old alloc/op   new alloc/op   delta
LastWriteWinsInsert/noPrevValue-8       177kB ± 4%     180kB ± 4%  +1.32%  (p=0.019 n=9+9)
LastWriteWinsInsert/withPrevValue-8     242kB ± 1%     244kB ± 0%  +0.80%  (p=0.000 n=9+8)

name                                 old allocs/op  new allocs/op  delta
LastWriteWinsInsert/noPrevValue-8       1.08k ± 1%     1.17k ± 1%  +8.72%  (p=0.000 n=8+8)
LastWriteWinsInsert/withPrevValue-8     1.63k ± 0%     1.73k ± 0%  +5.69%  (p=0.000 n=9+9)
```
We can avoid this by splitting `ProcessRow` into two methods.

File: pkg/ccl/crosscluster/logical/logical_replication_writer_processor.go [hunk 1]
```diff
@@ -531,43 +538,67 @@ type BatchHandler interface {
...
+	} else {
+		err = t.db.Txn(ctx, func(ctx context.Context, txn isql.Txn) error {
+			// TODO(ssd): For now, we SetOmitInRangefeeds to
+			// prevent the data from being emitted back to the source.
+			// However, I don't think we want to do this in the long run.
+			// Rather, we want to store the inbound cluster ID and store that
+			// in a way that allows us to choose to filter it out from or not.
+			// Doing it this way means that you can't choose to run CDC just from
+			// one side and not the other.
+			txn.KV().SetOmitInRangefeeds()
```
Comment (stevendanna) [inline pkg/ccl/crosscluster/logical/logical_replication_writer_processor.go:594]: Do we still need this here now that we are setting this in the session data?

**Review by stevendanna — APPROVED** [review]
Left a small question but overall this seems pretty good to me.  Makes sense that we wouldn't see an improvement in the micro-benchmark case. Will be interesting to see if this has any impact on a larger test.

**Review by yuzefovich — COMMENTED** [review]
There was initially some regression which I had to fight back in order to get "no difference in overhead" :smiley:

TFTR!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/126050#-:-O0ADDAW1-xClo57vD1b:b-g1v0gx)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @stevendanna)

___
*[`pkg/ccl/crosscluster/logical/logical_replication_writer_processor.go` line 591 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/126050#gh-1650348437:-O0ACnk77mLxxcG16JWU:b6iryer) ([raw file](https://github.com/cockroachdb/cockroach/blob/6914507a35add039761f03b6142ed4d03f627f85/pkg/ccl/crosscluster/logical/logical_replication_writer_processor.go#L591)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

Do we still need this here now that we are setting this in the session data?
</blockquote></details>

No, it's not needed, removed.

**Review by yuzefovich — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/126050)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained

___
*[`pkg/ccl/crosscluster/logical/logical_replication_writer_processor.go` line 591 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/126050#gh-1650348437:-O0H6d1F686Ovcz2WnkA:b-ahyhfu) ([raw file](https://github.com/cockroachdb/cockroach/blob/6914507a35add039761f03b6142ed4d03f627f85/pkg/ccl/crosscluster/logical/logical_replication_writer_processor.go#L591)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

No, it's not needed, removed.
</blockquote></details>

Turns out that we still need it - https://github.com/cockroachdb/cockroach/pull/126234.

**Pull request merged**